### PR TITLE
[spmd] Add reduction rule, use reduction rule in sum

### PR DESCRIPTION
**Pull request by wanchaol**
This PR enable reduction rule, add/rewrite sum to use the reduction
rule instead of manual op impl

### CODE REVIEW

**Review by fduwjj — APPROVED** [review]
lgtm.

File: spmd/tensor/ops/math_ops.py [hunk 1]
```diff
@@ -117,24 +117,37 @@ def einop_rule(
...
-        return DTensor(
-            local_sum, device_mesh=device_mesh, placements=self.placements
+def reduction_rule(op_schema: OpSchema) -> OutputSharding:
+    """
+    Propagate the sharding for reduction operations. Examples:
+        ij->i - sum on dim
+    """
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    # reduction op usually begin with a single tensor
+    input_spec = cast(PlacementSpec, op_schema.args_schema[0])
```
Comment (fduwjj) [inline spmd/tensor/ops/math_ops.py:127]: This is actually the spec of the input Distributed Tensor, right?

File: spmd/tensor/ops/math_ops.py [hunk 2]
```diff
@@ -117,24 +117,37 @@ def einop_rule(
...
+    if len(op_schema.args_schema) > 1 and isinstance(
+        op_schema.args_schema[1], (int, list)
+    ):
+        # this is usually a dim-based reduction op schema pattern, it
+        # might not be true for every op for other special cases, we
+        # need to specialize them as needed.
+        # TODO: add support for things like `torch.unique` where it
+        # does not follow the reduction op convention.
+        dim_list = as_list(op_schema.args_schema[1])
+        out_dimchars = input_chars.translate(
```
Comment (fduwjj) [inline spmd/tensor/ops/math_ops.py:139]: Interesting.. Python is magic!

File: spmd/tensor/ops/math_ops.py (hunk 1 above)
Comment (wanchaol) [reply spmd/tensor/ops/math_ops.py:127]: Yep

File: spmd/tensor/ops/math_ops.py (hunk 2 above)
Comment (wanchaol) [reply spmd/tensor/ops/math_ops.py:139]: translate seems super fast compare with other string replacement after some search, so i use this haha, magic indeed.

**Pull request merged**